compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/RegExpLexer.java

```diff
@@ -164,62 +116,123 @@ private STToken readTokenInReTerm() {
-                } else {
-                    startMode(ParserMode.RE_CHAR_SET);
-                }
                 return getRegExpSyntaxToken(SyntaxKind.OPEN_BRACKET_TOKEN);
-            // Handle "(" ["?" ReFlagsOnOff ":"] ReDisjunction ")".
+            case LexerTerminals.CLOSE_BRACKET:
+                this.reader.advance();
+                return getRegExpSyntaxToken(SyntaxKind.CLOSE_BRACKET_TOKEN);
+            case LexerTerminals.OPEN_BRACE:
+                this.reader.advance();
+                return getRegExpSyntaxToken(SyntaxKind.OPEN_BRACE_TOKEN);
+            case LexerTerminals.CLOSE_BRACE:
+                this.reader.advance();
+                return getRegExpSyntaxToken(SyntaxKind.CLOSE_BRACE_TOKEN);
+            // Start parsing capturing group ([? ReFlagsOnOff :] ReDisjunction).
             case LexerTerminals.OPEN_PARANTHESIS:
-                if (this.reader.peek(1) == LexerTerminals.QUESTION_MARK) {
-                    startMode(ParserMode.RE_FLAGS_START);
+                this.reader.advance();
+                if (peek() == LexerTerminals.QUESTION_MARK) {
+                    // Start a new parser mode to handle the flags.
+                    startMode(ParserMode.RE_FLAG_EXPRESSION);
                 } else {
-                    startMode(ParserMode.RE_CAPTURING_GROUP_RE_DISJUNCTION);
+                    startMode(ParserMode.RE_DISJUNCTION);
                 }
-                this.reader.advance();
                 return getRegExpSyntaxToken(SyntaxKind.OPEN_PAREN_TOKEN);
             case LexerTerminals.CLOSE_PARANTHESIS:
                 this.reader.advance();
-                if (this.mode == ParserMode.RE_CAPTURING_GROUP_RE_DISJUNCTION) {
-                    switchMode(ParserMode.RE_QUANTIFIER);
-                    return getRegExpSyntaxToken(SyntaxKind.CLOSE_PAREN_TOKEN);
+                // modeStack.size() is > 1 if the close parenthesis is found in a capturing group.
+                // If not it's an invalid token.
+                if (this.modeStack.size() > 1) {
+                    endMode();
                 }
-                reportLexerError(DiagnosticErrorCode.ERROR_INVALID_TOKEN_IN_REG_EXP);
-                return getRegExpText(SyntaxKind.RE_CHAR);
+                return getRegExpSyntaxToken(SyntaxKind.CLOSE_PAREN_TOKEN);
+            case LexerTerminals.COMMA:
+                this.reader.advance();
+                return getRegExpSyntaxToken(SyntaxKind.COMMA_TOKEN);
             default:
-                // Handle ReLiteralChar.
                 this.reader.advance();
-                if (!isReSyntaxChar(nextChar)) {
-                    startMode(ParserMode.RE_QUANTIFIER);
+                if (isDigit(nextChar)) {
+                    return getRegExpText(SyntaxKind.DIGIT);
+                }
+                return getRegExpText(SyntaxKind.RE_LITERAL_CHAR);
+        }
+    }
+
+    private STToken processReEscape() {
+        if (reader.isEOF()) {
```

Comment:
this is redundant as we reach here for the nextChar being `LexerTerminals.BACKSLASH:`